util/mdbook/move_md.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+'''Moves a markdown file, recalculating the relative links for the new location
+
+KNOWN_ISSUES
+- If you move a file to the same dir an existing relative link points to, the path
+  won't be canonicalized (you will get '../<dir>' rather than '.')
+- Doesn't check if link to anchor is correct, just check that file exists
+- Would be nice to re-render the markdown straight from the AST, but I found other
+  formatting changes when trying this.
+'''
+
+import validators
+from markdown_it import MarkdownIt
+from markdown_it.tree import SyntaxTreeNode
+
+import argparse
+import pathlib
+from pathlib import Path
+
+import logging
+logging.basicConfig()
+logging.getLogger().setLevel(logging.DEBUG)  # set root-logger to most-permissive
+logger = logging.getLogger(__name__)
+
+logging.getLogger('markdown_it').setLevel(logging.WARNING)
+
+
+# https://stackoverflow.com/a/71874881
+def relative(target: Path, origin: Path) -> Path:
+    """Return the path of target relative to origin.
+    i.e. --> pathlib equivalent of os.path.relpath'
+    """
+    try:
+        return Path(target).resolve().relative_to(Path(origin).resolve())
+    except ValueError:  # target does not start with origin
+        # recursion with origin (eventually origin is root so try will succeed)
+        return Path('..').joinpath(relative(target, Path(origin).parent))
+
+
+def do_update(l: str, old_path: pathlib.Path) -> bool:
+    """Do some checking on the relative link in the existing file.
+
+    Return 'False' if we don't need to recalculate it.
+    """
+
+    # Don't update link if it is a URL
+    if validators.url(l):
+        logger.debug(f"IsUrl : {l}")
+        return False
+
+    # Strip fragment/anchor if present
+    l_path, l_frag = None, None
+    if "#" in l:
+        l_path, l_frag = l.split("#")
+    else:
+        l_path, l_frag = l, None
+    if not l_path:
+        # Must be a local anchor in the same file
+        # We don't need to update this one.
+        return False
+
+    oldlink_target = (old_path.parents[0]/Path(l_path)).resolve()
+    if not oldlink_target.exists():
+        logger.warning("Link in old_path points to non-existent file.")
+        logger.warning(f"  OLD_PATH       : {old_path}")
+        logger.warning(f"  REL_LINK       : {l}")
+        logger.warning(f"  TARGET   (BAD) : {oldlink_target}")
+
+    return True
+
+
+def move_md(old_path: pathlib.Path, new_path: pathlib.Path) -> None:
+    '''Main link rewriter'''
+    if new_path.suffix != '.md':
```

Comment:
Are the fallthrough intentional? I don't think that logger.error throws an exception.

Reply:
I will change to raise an exception here.